src/initializer.tsx

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2021 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import Olm from "@matrix-org/olm";
+import olmWasmPath from "@matrix-org/olm/olm.wasm?url";
+import { Integrations } from "@sentry/tracing";
+import i18n from "i18next";
+import { initReactI18next } from "react-i18next";
+import LanguageDetector from "i18next-browser-languagedetector";
+import Backend from "i18next-http-backend";
+import * as Sentry from "@sentry/react";
+
+import { getUrlParams } from "./UrlParams";
+import { Config } from "./config/Config";
+import { DEFAULT_CONFIG } from "./config/ConfigOptions";
+
+enum LoadState {
+  None,
+  Loading,
+  Loaded,
+}
+
+class DependencyLoadStates {
+  olm: LoadState = LoadState.None;
+  config: LoadState = LoadState.None;
+  sentry: LoadState = LoadState.None;
+  i18n: LoadState = LoadState.None;
+
+  allDepsAreLoaded() {
+    const notLoadedDeps = Object.values(this).filter(
+      (s) => s !== LoadState.Loaded
+    );
+    return notLoadedDeps.length === 0;
+  }
+}
+
+export class Initializer {
+  private static _instance: Initializer;
+
+  public static init(): Promise<void> | null {
+    if (Initializer?._instance?.initPromise) {
+      return null;
```

Comment:
null?

Reply:
otherwise we will attach multiple `then`s to the promise.
which would call `setLoaded(true);` multiple times. Not an issue but if someone adds functionality it might be confusing why its called multiple times? Or would that be expected?
```
useEffect(() => {
    Initializer.init()?.then(() => {
      setLoaded(true);
    });
  });
  ```

Reply:
Looks like its fine:
![image](https://user-images.githubusercontent.com/16718859/198883308-7bb57cbb-eea7-4c81-9316-515b582940fb.png)

Reply:
Ah, yeah, right. Makes sense